File: samples/CslaNet/vb/ProjectTrackervb/PTWin/RolesEdit.vb
Imports ProjectTracker.Library.Admin

Public Class RolesEdit
  Inherits WinPart

  Private _roles As Admin.Roles

  Private Sub RolesEdit_Load( _
    ByVal sender As System.Object, ByVal e As System.EventArgs) _
    Handles MyBase.Load

    Try
      _roles = Admin.Roles.GetRoles

    Catch ex As Csla.DataPortalException
      MessageBox.Show(ex.BusinessException.ToString, _
        "Error loading", MessageBoxButtons.OK, _
        MessageBoxIcon.Exclamation)

    Catch ex As Exception
      MessageBox.Show(ex.ToString, _
        "Error loading", MessageBoxButtons.OK, _
        MessageBoxIcon.Exclamation)
    End Try

    If _roles IsNot Nothing Then
      Me.RolesBindingSource.DataSource = _roles
    End If

  End Sub

  Protected Overrides Function GetIdValue() As Object

    Return "Edit Roles"

  End Function

  Private Sub SaveButton_Click( _
    ByVal sender As System.Object, ByVal e As System.EventArgs) _
    Handles SaveButton.Click

    ' stop the flow of events
    Me.RolesBindingSource.RaiseListChangedEvents = False
    ' commit edits in memory
    UnbindBindingSource(Me.RolesBindingSource, True, True)
    Try
      _roles = _roles.Save
      Me.Close()

    Catch ex As Csla.DataPortalException
      MessageBox.Show(ex.BusinessException.ToString, _
        "Error saving", MessageBoxButtons.OK, _
        MessageBoxIcon.Exclamation)

    Catch ex As Exception
      MessageBox.Show(ex.ToString, _
        "Error saving", MessageBoxButtons.OK, _
        MessageBoxIcon.Exclamation)

    Finally
      Me.RolesBindingSource.DataSource = _roles

      Me.RolesBindingSource.RaiseListChangedEvents = True
      Me.RolesBindingSource.ResetBindings(False)
    End Try

  End Sub

  Private Sub CancelButton_Click( _
    ByVal sender As System.Object, ByVal e As System.EventArgs) _
    Handles CancelButton.Click

    Me.Close()

  End Sub

  Private Sub RolesEdit_CurrentPrincipalChanged( _
    ByVal sender As Object, ByVal e As System.EventArgs) _
    Handles Me.CurrentPrincipalChanged

    If Not Roles.CanEditObject Then
      Me.Close()
    End If

  End Sub

End Class

File: Samples/ProjectTracker/ProjectTracker.Ui.WinForms/RolesEdit.cs
using System;
using System.Windows.Forms;
using ProjectTracker.Library;
using ProjectTracker.Library.Admin;

namespace PTWin
{
  public partial class RolesEdit : WinPart
  {
    private RoleEditBindingList _roles;

    public RolesEdit()
    {
      InitializeComponent();
    }

    private void RolesEdit_Load(object sender, EventArgs e)
    {
      try
      {
        _roles = RoleEditBindingList.GetRoles();
      }
      catch (Csla.DataPortalException ex)
      {
        MessageBox.Show(ex.BusinessException.ToString(),
          "Data load error", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
      }
      catch (Exception ex)
      {
        MessageBox.Show(ex.ToString(),
          "Data load error", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
      }

      if (_roles != null)
        this.rolesBindingSource.DataSource = _roles;
    }

    protected internal override object GetIdValue()
    {
      return "Edit Roles";
    }

    private void SaveButton_Click(object sender, EventArgs e)
    {
      // stop the flow of events
      this.rolesBindingSource.RaiseListChangedEvents = false;

      // commit edits in memory and unbind
      UnbindBindingSource(this.rolesBindingSource, true, true);

      try
      {
        _roles = _roles.Save();
        RoleList.InvalidateCache();
        RoleList.CacheList();
        this.Close();
      }
      catch (Csla.DataPortalException ex)
      {
        MessageBox.Show(ex.BusinessException.ToString(),
          "Error saving", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
      }
      catch (Exception ex)
      {
        MessageBox.Show(ex.ToString(),
          "Error saving", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
      }
      finally
      {
        this.rolesBindingSource.DataSource = _roles;

        this.rolesBindingSource.RaiseListChangedEvents = true;
        this.rolesBindingSource.ResetBindings(false);
      }
    }

    private void CancelButton_Click(object sender, EventArgs e)
    {
      this.Close();
    }

    private void RolesEdit_CurrentPrincipalChanged(
      object sender, EventArgs e)
    {
      if (!Csla.Rules.BusinessRules.HasPermission(Csla.Rules.AuthorizationActions.EditObject, typeof(RoleEditBindingList)))
        this.Close();
    }
  }
}